examples/passing/LetPattern.purs

```diff
@@ -0,0 +1,128 @@
+module Main where
+
+import Prelude
+import Partial.Unsafe (unsafePartial)
+import Control.Monad.Eff (Eff)
+import Control.Monad.Eff.Console (CONSOLE, log)
+import Test.Assert (ASSERT, assert')
+
+patternSimple :: Boolean
+patternSimple =
+  let (x) = 25252
+  in
+   x == 25252
+
+patternDoSimple :: forall e. Eff e Boolean
+patternDoSimple = do
+  let (x) = 25252
+  pure $ x == 25252
+
+newtype X = X Int
+
+patternNewtype :: Boolean
+patternNewtype =
+  let (X a) = X 123
+  in
+   a == 123
+
+patternDoNewtype :: forall e. Eff e Boolean
+patternDoNewtype = do
+  let (X a) = X 123
+  pure $ a == 123
+
+data Y = Y Int String Boolean
+
+patternData :: Boolean
+patternData =
+  let (Y a b c) = Y 456 "hello, world" false
+  in
+   a == 456 && b == "hello, world" && not c
+
+patternDataIgnored :: Boolean
+patternDataIgnored =
+  let (Y _ x _) = Y 789 "world, hello" true
+  in
+   x == "world, hello"
+
+patternDoData :: forall e. Eff e Boolean
+patternDoData = do
+  let (Y a b c) = Y 456 "hello, world" false
+  pure $ a == 456 && b == "hello, world" && not c
+
+patternDoDataIgnored :: forall e. Eff e Boolean
+patternDoDataIgnored = do
+  let (Y _ x _) = Y 789 "world, hello" true
+  pure $ x == "world, hello"
+
+patternArray :: Boolean
+patternArray = unsafePartial $
+  let [a, b] = [1, 2]
+  in
+   a == 1 && b == 2
+
+patternDoArray :: forall e. Eff e Boolean
+patternDoArray = unsafePartial do
+  let [a, b] = [1, 2]
+  pure $ a == 1 && b == 2
+
+patternMultiple :: Boolean
```

Comment:
Well, we would need to run the binding groups pass before the `let` pass. It would be quite a bit more work. There are also issues related to code gen, as described on the original issue, so maybe we should leave that for later.

Any thoughts, anyone else?